Add Artifact class and WorkflowRun.get_artifacts

yixinguo2 opened the pull request:
Add ability to get artifacts information and download build artifacts from workflow runs.
- Added Artifact class
- Added WorkflowRun.get_artifacts
- Added WorkflowRun.test_get_artifacts
- Added tests/Artifacts.py

yixinguo2 commented on github/Artifact.py:
```diff
@@ -0,0 +1,157 @@
+    def __repr__(self):
+        return self.get__repr__({"id": self._id.value, "url": self._url.value})
+
+    def get_file_content(self):
```
This is the only method that does not have a test mainly because I didn't know how to go through the GithubObject connection to get the file content and used requests instead.

bauergeorg commented:
I wait for final implementation. But this works for me. 

 ```python
 def get_artifact(self, run):
        ''' Get download url for one single download run of a artifact.

        Return:
            info (dict or list): e.g. {'name': 'sandbox-package1, 'url': 'https://pipelines.actions.githubusercontent.com/ddYkl%3D'}
        '''
        import json

        # get artifact content
        status, artifactHeader, artifact = run._requester.requestJson("GET", run.artifacts_url)
        artifact_content = json.loads(artifact)

        # get archive url
        ret =[]
        for i in range(artifact_content['total_count']):
            name = artifact_content['artifacts'][0]['name']
            url = artifact_content['artifacts'][0]['archive_download_url']

            # generates a download for one time
            #out = run._requester.requestMultipart(verb = "GET", url = url)
            _, responseHeaders, output = run._requester.requestMultipart(verb = "GET", url = url, parameters=None, headers=None, input=None, cnx=run._requester._Requester__customConnection(url))
            # manage return value
            info_dict = {'name': name, 'url': responseHeaders['location']}
            ret.append(info_dict)

        # manage return
        if len(ret) < 1:
            return None
        elif len(ret) == 1:
            return ret[0]
        else:
            return ret
```

To get file content:
```python
    def get_zip_content(self, download_url):
        ''' Returns zip file content as a list '''
        import requests
        from io import BytesIO
        from zipfile import ZipFile

        url = requests.get(download_url)
        zipfile = ZipFile(BytesIO(url.content))
        zip_content = zipfile.namelist()

        return zip_content
```

yixinguo2 commented:
> I wait for final implementation. But this works for me.

I took your suggestion to mean removing the Artifact class and just add the method to Workflow to return the name and url map since it takes in a run object. It made the most sense to me. I've also updated tests for Workflow and fixed a bug where `name = artifact_content['artifacts'][0]['name']` should be `name = artifact_content['artifacts'][i]['name']`. Please take a look at the updated code.

yixinguo2 commented:
@bauergeorg could you approve the workflow above? I'm a first-time contributor.

bauergeorg commented:
@yixinguo2 sorry I have no access rights.

yixinguo2 commented:
@s-t-e-v-e-n-k could you approve the workflow? I'm a first-time contributor.

bauergeorg commented:
Could you reopen this pull request?

yixinguo2 commented:
> Could you reopen this pull request?

@bauergeorg sorry I was switching jobs when this PR was still open. I have created another PR to continue the submission process https://github.com/PyGithub/PyGithub/pull/2245

bauergeorg commented:
No problem. Thanks. Enjoy your new business.
